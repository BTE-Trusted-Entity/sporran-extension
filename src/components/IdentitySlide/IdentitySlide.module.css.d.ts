export const __esModule: true;
export const container: string;
export const nameLine: string;
export const centeredNameLine: string;
export const name: string;
export const form: string;
export const input: string;
export const save: string;
export const cancel: string;
export const add: string;
