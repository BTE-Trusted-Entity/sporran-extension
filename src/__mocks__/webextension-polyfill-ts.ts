// src/__mocks__/webextension-polyfill-ts
// Update this file to include any mocks for the `webextension-polyfill-ts` package
// This is used to mock these values for Storybook so you can develop your components
// outside the Web Extension environment provided by a compatible browser

import { setupGetMessageShim } from 'chrome-extension-i18n-shim';

import messagesEN from '../static/_locales/en/messages.json';

export const browser = {
  tabs: {
    executeScript(
      currentTabId: number /* eslint-disable-line @typescript-eslint/no-unused-vars */,
      details: unknown /* eslint-disable-line @typescript-eslint/no-unused-vars */,
    ): Promise<{ done: boolean }> {
      return Promise.resolve({ done: true });
    },
  },
  i18n: {
    getMessage: setupGetMessageShim(messagesEN),
    getUILanguage(): string {
      return document.documentElement.lang || 'en';
    },
  },
  storage: {
    local: {
      async get(): Promise<unknown> {
        return {
          authorizedDApps: {
            'example.com': false,
            'examplesuperlong.dfghrtydfghklyuihjwertyxcvbn3456rty56782345678yudfgghtyuwertyudfghsdfgh-domain.org':
              true,
            'example.org': true,
          },
        };
      },
      async set(): Promise<void> {
        // dummy
      },
    },
  },
  runtime: {
    async sendMessage(): Promise<void> {
      // dummy
    },
    onMessage: {
      addListener(): void {
        // dummy
      },
      removeListener(): void {
        // dummy
      },
    },
  },
};

export interface Tabs {
  Tab: {
    id: number;
  };
}
