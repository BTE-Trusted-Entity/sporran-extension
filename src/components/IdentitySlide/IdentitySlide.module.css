.container {
  width: 100%;
}

.nameLine {
  composes: identityText from '../Typography/Typography.module.css';
  position: relative;
  padding-left: 2.125rem;
  height: 1.5rem;
  display: flex;
  align-items: center;
  justify-content: center;
  margin-top: 0.375rem;
}

.centeredNameLine {
  composes: nameLine;
  padding-left: 0;
}

.name {
  overflow: hidden;
  white-space: nowrap;
  text-overflow: ellipsis;
}

.form {
  display: flex;
  align-items: center;
  justify-content: center;
}

.input {
  composes: inputName from '../Input/Input.module.css';
  width: calc(232rem / 16);
  margin: 1rem 0;
}

.save {
  composes: buttonCheckmark from '../Button/Button.module.css';
}

.cancel {
  composes: buttonCross from '../Button/Button.module.css';
}

.add {
  composes: avatar from '../../components/Avatar/Avatar.module.css';
  background: rgb(var(--color-black-rgb) / 30%) url('../images/plus-circle.svg')
    no-repeat center/100%;
  border-radius: 50%;
  margin-bottom: 1.625rem;
}
