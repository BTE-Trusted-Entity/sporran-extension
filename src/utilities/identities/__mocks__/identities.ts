export const decryptIdentity = jest.fn();

export const saveIdentity = jest.fn();
