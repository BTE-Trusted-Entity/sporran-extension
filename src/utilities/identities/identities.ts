import { useContext } from 'react';
import { mutate } from 'swr';
import {
  blake2AsU8a,
  keyExtractPath,
  keyFromPath,
  mnemonicToMiniSecret,
  sr25519PairFromSeed,
} from '@polkadot/util-crypto';
import {
  Did,
  DidDocument,
  DidResourceUri,
  DidUri,
  EncryptRequestData,
  IEncryptedMessage,
  KiltAddress,
  KiltEncryptionKeypair,
  KiltKeyringPair,
  Message,
  MessageBody,
  SignCallback,
  SignRequestData,
  Utils,
} from '@kiltprotocol/sdk-js';
import { map, max, memoize } from 'lodash-es';

import {
  loadEncrypted,
  saveEncrypted,
} from '../storageEncryption/storageEncryption';

import { getDidDocument, getDidEncryptionKey, parseDidUri } from '../did/did';
import { storage } from '../storage/storage';
import { useSwrDataOrThrow } from '../useSwrDataOrThrow/useSwrDataOrThrow';

import { IdentitiesContext, IdentitiesContextType } from './IdentitiesContext';
import { getIdentities, IDENTITIES_KEY } from './getIdentities';

import { Identity } from './types';

export { Identity, IdentitiesMap } from './types';

const CURRENT_IDENTITY_KEY = 'currentIdentity';

export const NEW: Identity = {
  address: 'NEW' as KiltAddress,
  did: '' as DidUri,
  name: '',
  index: -1,
};

export function isNew(identity: Identity): boolean {
  return identity === NEW;
}

export function useIdentities(): IdentitiesContextType {
  return useContext(IdentitiesContext);
}

async function getCurrentIdentity(): Promise<string | null> {
  const identities = await getIdentities();
  const stored = await storage.get(CURRENT_IDENTITY_KEY);
  const current = stored[CURRENT_IDENTITY_KEY];

  if (identities[current]) {
    return current;
  }

  const firstIdentity = Object.values(identities)[0];
  if (!firstIdentity) {
    return null;
  }

  await setCurrentIdentity(firstIdentity.address);

  return firstIdentity.address;
}

export async function setCurrentIdentity(address: string): Promise<void> {
  const oldAddress = await storage.get(CURRENT_IDENTITY_KEY);
  if (address === oldAddress[CURRENT_IDENTITY_KEY]) {
    return;
  }
  await storage.set({ [CURRENT_IDENTITY_KEY]: address });
  await mutate(['getCurrentIdentity', CURRENT_IDENTITY_KEY]);
}

export async function saveIdentity(identity: Identity): Promise<void> {
  const identities = await getIdentities();
  identities[identity.address] = identity;
  await storage.set({ [IDENTITIES_KEY]: identities });
  await mutate(IDENTITIES_KEY);
}

export async function removeIdentity(identity: Identity): Promise<void> {
  const identities = await getIdentities();
  delete identities[identity.address];

  await storage.set({ [IDENTITIES_KEY]: identities });
  await storage.remove(identity.address);

  const firstIdentity = Object.values(identities)[0];
  await setCurrentIdentity(firstIdentity?.address);

  await mutate(IDENTITIES_KEY);
}

interface IdentityDidCrypto {
  didDocument: DidDocument;
  keypair: KiltKeyringPair;
  authenticationKey: KiltKeyringPair;
  sign: SignCallback;
  encrypt: (
    messageBody: MessageBody,
    dAppDidDocument: DidDocument,
  ) => Promise<IEncryptedMessage>;
}

export function deriveAuthenticationKey(seed: Uint8Array) {
  const baseKey = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  return baseKey.derive('//did//0') as typeof baseKey;
}

export function deriveEncryptionKeyFromSeed(
  seed: Uint8Array,
): KiltEncryptionKeypair {
  const keypair = sr25519PairFromSeed(seed);
  const { path } = keyExtractPath('//did//keyAgreement//0');
  const { secretKey } = keyFromPath(keypair, path, 'sr25519');
  return Utils.Crypto.makeEncryptionKeypairFromSeed(blake2AsU8a(secretKey));
}

export function deriveAttestationKeyFromSeed(seed: Uint8Array) {
  const baseKey = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  return baseKey.derive('//did//assertion//0') as typeof baseKey;
}

function deriveEncryptionKeyLegacy(seed: Uint8Array) {
  const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const encryptionKeyringPair = keypair.derive('//did//keyAgreement//0');

  const secret = encryptionKeyringPair
    .encryptMessage(
      new Uint8Array(24).fill(0),
      new Uint8Array(24).fill(0),
      new Uint8Array(24).fill(0),
    )
    .slice(24); // first 24 bytes are the nonce

  return Utils.Crypto.makeEncryptionKeypairFromSeed(secret);
}

async function fixLightDidIssues(seed: Uint8Array) {
  const identities = await getIdentities();
  const { address } = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const identity = identities[address];

  if (!identity) {
    // could be the Alice identity
    return;
  }

  const parsed = identity.did && Did.parse(identity.did);
  if (parsed && parsed.type !== 'light') {
    return;
  }

  try {
    if (!identity.did) {
      // DID was deactivated, no action needed.
      return;
    }

    // If this light DID was created and stored using SDK@0.24.0 then its keys are serialized using base64,
    // resulting in an invalid URI, so resolving would throw an exception.
    const document = await getDidDocument(identity.did);

    // Another issue we see is the light DIDs without key agreement keys, need to regenerate them as well
    const encryptionKey = getDidEncryptionKey(document);

    // This public key also means the DID needs to be regenerated
    const troubleKey =
      '0xf2c90875e0630bd1700412341e5e9339a57d2fefdbba08de1cac8db5b4145f6e';
    if (Utils.Crypto.u8aToHex(encryptionKey.publicKey) === troubleKey) {
      throw new Error();
    }
  } catch {
    // We re-create the invalid DID from scratch and update its URI in the identity.
    const { uri } = getLightDidFromSeed(seed);
    await saveIdentity({ ...identity, did: uri });
  }
}

export async function getIdentityCryptoFromSeed(
  seed: Uint8Array,
  legacy?: boolean,
): Promise<IdentityDidCrypto> {
  await fixLightDidIssues(seed);

  const authenticationKey = deriveAuthenticationKey(seed);
  const encryptionKey = legacy
    ? deriveEncryptionKeyLegacy(seed)
    : deriveEncryptionKeyFromSeed(seed);

  const keypair = Utils.Crypto.makeKeypairFromSeed(seed, 'sr25519');
  const identities = await getIdentities();
  const did = getIdentityDid(identities[keypair.address]);

  const didDocument = await getDidDocument(did);

  async function encryptCallback({ data, peerPublicKey }: EncryptRequestData) {
    const { box, nonce } = Utils.Crypto.encryptAsymmetric(
      data,
      peerPublicKey,
      encryptionKey.secretKey,
    );
    const keyUri = `${didDocument.uri}${getDidEncryptionKey(didDocument).id}`;
    return {
      nonce,
      data: box,
      keyUri: keyUri as DidResourceUri,
    };
  }

  async function sign({ data, keyRelationship }: SignRequestData) {
    if (keyRelationship !== 'authentication') {
      throw new Error(
        'Only key relationship "authentication" is supported for signing',
      );
    }

    const signature = authenticationKey.sign(data, { withType: false });
    const keyUri =
      `${didDocument.uri}${didDocument.authentication[0].id}` as DidResourceUri;
    const keyType = authenticationKey.type;

    return {
      signature,
      keyUri,
      keyType,
    };
  }

  async function encrypt(
    messageBody: MessageBody,
    dAppDidDocument: DidDocument,
  ): Promise<IEncryptedMessage> {
    const message = Message.fromBody(
      messageBody,
      didDocument.uri,
      dAppDidDocument.uri,
    );

    return Message.encrypt(
      message,
      encryptCallback,
      `${dAppDidDocument.uri}${getDidEncryptionKey(dAppDidDocument).id}`,
    );
  }

  return {
    didDocument,
    keypair,
    authenticationKey,
    sign,
    encrypt,
  };
}

export async function encryptIdentity(
  backupPhrase: string,
  password: string,
): Promise<KiltAddress> {
  const seed = mnemonicToMiniSecret(backupPhrase);
  const { address } = Utils.Crypto.makeKeypairFromUri(backupPhrase, 'sr25519');
  await saveEncrypted(address, password, seed);
  return address;
}

export function getLightDidFromSeed(seed: Uint8Array): DidDocument {
  return Did.createLightDidDocument({
    authentication: [deriveAuthenticationKey(seed)],
    keyAgreement: [deriveEncryptionKeyFromSeed(seed)],
  });
}

async function getIdentityName(): Promise<{ name: string; index: number }> {
  const identities = await getIdentities();
  const largestIndex = max(map(identities, 'index')) || 0;

  const index = 1 + largestIndex;

  return { name: `KILT Identity ${index}`, index };
}

export async function createIdentity(
  backupPhrase: string,
  password: string,
): Promise<Identity> {
  const address = await encryptIdentity(backupPhrase, password);

  const seed = mnemonicToMiniSecret(backupPhrase);

  const { uri } = getLightDidFromSeed(seed);

  const { name, index } = await getIdentityName();

  const identity = { name, address, did: uri, index };
  await saveIdentity(identity);

  return identity;
}

export async function importIdentity(
  backupPhrase: string,
  password: string,
): Promise<Identity> {
  const address = await encryptIdentity(backupPhrase, password);

  const seed = mnemonicToMiniSecret(backupPhrase);

  const lightDidDocument = getLightDidFromSeed(seed);
  const lightDid = lightDidDocument.uri;

  const resolved = await Did.resolve(lightDid);
  const fullDid = resolved?.metadata?.canonicalId;

  const deactivated = resolved?.metadata?.deactivated;
  const did = deactivated ? undefined : fullDid || lightDid;
  const deletedDid = deactivated ? Did.getFullDidUri(lightDid) : undefined;

  const { name, index } = await getIdentityName();

  const identity = {
    name,
    address,
    index,
    ...(!deactivated && { did }),
    ...(deactivated && { deletedDid }),
  };
  await saveIdentity(identity);

  return identity;
}

export async function decryptIdentity(
  address: string,
  password: string,
): Promise<Uint8Array> {
  const seed = await loadEncrypted(address, password);
  return new Uint8Array(seed);
}

/** Ensure that local information about the DID type matches stored on blockchain
 * even if an error occurred while asynchronous update was running */
async function syncDidStateWithBlockchain(address: string | null | undefined) {
  if (!address) {
    return;
  }

  const identities = await getIdentities();
  const identity = identities[address];

  if (!identity.did) {
    // could be a legacy identity without DID
    return;
  }

  const { fullDid, type } = parseDidUri(identity.did);
  const wasOnChain = type === 'full';

  const resolved = await Did.resolve(identity.did);
  const isOnChain = wasOnChain
    ? Boolean(resolved && resolved.metadata && !resolved.metadata.deactivated)
    : Boolean(resolved && resolved.metadata && resolved.metadata.canonicalId);

  if (wasOnChain && !isOnChain) {
    await saveIdentity({
      ...identity,
      did: undefined,
      deletedDid: fullDid,
    });
  }
  if (!wasOnChain && isOnChain) {
    await saveIdentity({ ...identity, did: fullDid });
  }
}

/** Memoized function will run only once per identity while the popup is open, do not use from backend */
const noAwaitUpdateCachedDidStateOnce = memoize(syncDidStateWithBlockchain);

export function useCurrentIdentity(): string | null | undefined {
  const data = useSwrDataOrThrow(
    CURRENT_IDENTITY_KEY,
    getCurrentIdentity,
    'getCurrentIdentity',
  );

  noAwaitUpdateCachedDidStateOnce(data);

  return data;
}

export function getIdentityDid({ did }: Identity): DidUri {
  if (!did) {
    throw new Error('DID is deleted and unusable');
  }
  return did;
}
