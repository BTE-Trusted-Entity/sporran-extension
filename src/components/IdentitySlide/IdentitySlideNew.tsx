import { browser } from 'webextension-polyfill-ts';
import { Link } from 'react-router-dom';

import * as styles from './IdentitySlide.module.css';

import { paths } from '../../views/paths';

export function IdentitySlideNew(): JSX.Element {
  const t = browser.i18n.getMessage;

  return (
    <section>
      <Link
        to={paths.identity.add}
        className={styles.add}
        aria-label={t('component_IdentitySlideNew_title')}
        title={t('component_IdentitySlideNew_title')}
      />
    </section>
  );
}
