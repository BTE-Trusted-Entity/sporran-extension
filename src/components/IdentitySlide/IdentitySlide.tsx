import { FormEvent, useCallback } from 'react';
import { browser } from 'webextension-polyfill-ts';

import * as styles from './IdentitySlide.module.css';

import { Avatar } from '../Avatar/Avatar';
import { IdentityOptions } from '../IdentityOptions/IdentityOptions';

import { Identity, saveIdentity } from '../../utilities/identities/identities';
import { useBooleanState } from '../../utilities/useBooleanState/useBooleanState';

interface Props {
  identity: Identity;
  options?: boolean;
}

export function IdentitySlide({
  identity,
  options = false,
}: Props): JSX.Element {
  const t = browser.i18n.getMessage;

  const editing = useBooleanState();

  const handleSubmit = useCallback(
    async (event: FormEvent) => {
      event.preventDefault();

      const target = event.target as unknown as {
        elements: Record<string, HTMLInputElement>;
      };

      const name = target.elements.name.value;
      await saveIdentity({
        ...identity,
        name,
      });

      editing.off();
    },
    [editing, identity],
  );

  return (
    <section className={styles.container}>
      <Avatar identity={identity} />
      {!editing.current ? (
        <div className={options ? styles.nameLine : styles.centeredNameLine}>
          <span className={styles.name}>{identity.name}</span>
          {options && (
            <IdentityOptions identity={identity} onEdit={editing.on} />
          )}
        </div>
      ) : (
        <form className={styles.form} onSubmit={handleSubmit}>
          <input
            className={styles.input}
            name="name"
            required
            aria-label={t('component_IdentitySlide_name')}
            placeholder={t('component_IdentitySlide_name')}
            defaultValue={identity.name}
            autoComplete="off"
            autoFocus
          />
          <button
            className={styles.save}
            type="submit"
            aria-label={t('common_action_save')}
            title={t('common_action_save')}
          />
          <button
            className={styles.cancel}
            type="button"
            onClick={editing.off}
            aria-label={t('common_action_cancel')}
            title={t('common_action_cancel')}
          />
        </form>
      )}
    </section>
  );
}
