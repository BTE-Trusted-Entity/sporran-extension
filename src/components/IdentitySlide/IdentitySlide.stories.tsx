import { Meta } from '@storybook/react';

import { identitiesMock as identities } from '../../utilities/identities/IdentitiesProvider.mock';

import { IdentitySlide } from './IdentitySlide';
import { IdentitySlideNew } from './IdentitySlideNew';

export default {
  title: 'Components/IdentitySlide',
  component: IdentitySlide,
  decorators: [(story) => <div style={{ textAlign: 'center' }}>{story()}</div>],
} as Meta;

export function Template(): JSX.Element {
  return (
    <IdentitySlide
      identity={identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1']}
    />
  );
}

export function LongName(): JSX.Element {
  return (
    <IdentitySlide
      identity={identities['4pUVoTJ69JMuapNducHJPU68nGkQXB7R9xAWY9dmvUh42653']}
    />
  );
}

export function New(): JSX.Element {
  return <IdentitySlideNew />;
}

export function withFullDid(): JSX.Element {
  return (
    <IdentitySlide
      identity={identities['4pNXuxPWhMxhRctgB4qd3MkRt2Sxp7Y7sxrApVCVXCEcdQMo']}
    />
  );
}
