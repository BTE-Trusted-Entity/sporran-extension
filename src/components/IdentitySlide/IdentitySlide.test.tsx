import userEvent from '@testing-library/user-event';
import { browser } from 'webextension-polyfill-ts';

import {
  identitiesMock as identities,
  render,
  screen,
} from '../../testing/testing';
import { saveIdentity } from '../../utilities/identities/identities';

import { IdentitySlide } from './IdentitySlide';
import { IdentitySlideNew } from './IdentitySlideNew';

jest.mock('../../utilities/identities/identities');
jest.spyOn(browser.runtime, 'sendMessage');

const identity = identities['4tDjyLy2gESkLzvaLnpbn7N61VgnwAhqnTHsPPFAwaZjGwP1'];

describe('IdentitySlide', () => {
  it('should render', async () => {
    const { container } = render(<IdentitySlide identity={identity} />);
    expect(container).toMatchSnapshot();
  });

  it('should enable editing the identity name', async () => {
    const { container } = render(<IdentitySlide identity={identity} options />);

    await userEvent.click(await screen.findByLabelText('Identity options'));
    await userEvent.click(
      await screen.findByRole('menuitem', { name: 'Edit Identity Name' }),
    );
    await userEvent.type(
      await screen.findByLabelText('Identity name:'),
      ' Foo',
    );

    expect(container).toMatchSnapshot();

    const saveButton = await screen.findByRole('button', { name: 'Save' });
    await userEvent.click(saveButton);

    expect(saveIdentity).toHaveBeenCalledWith({
      name: 'Light DID Identity Foo',
      address: identity.address,
      did: identity.did,
      index: 1,
    });

    expect(
      await screen.findByLabelText('Identity options'),
    ).toBeInTheDocument();

    expect(container).toMatchSnapshot();
  });
});

describe('IdentitySlideNew', () => {
  it('should render', async () => {
    const { container } = render(<IdentitySlideNew />);

    expect(container).toMatchSnapshot();
  });
});
