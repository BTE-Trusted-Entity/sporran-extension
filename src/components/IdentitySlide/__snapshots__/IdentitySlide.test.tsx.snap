// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IdentitySlide should enable editing the identity name 1`] = `
<div>
  <section
    class="container"
  >
    <div
      class="avatar"
    >
      Identicon
    </div>
    <form
      class="form"
    >
      <input
        aria-label="Identity name:"
        autocomplete="off"
        class="input"
        name="name"
        placeholder="Identity name:"
        required=""
        value="Light DID Identity"
      />
      <button
        aria-label="Save"
        class="save"
        title="Save"
        type="submit"
      />
      <button
        aria-label="Cancel"
        class="cancel"
        title="Cancel"
        type="button"
      />
    </form>
  </section>
</div>
`;

exports[`IdentitySlide should enable editing the identity name 2`] = `
<div>
  <section
    class="container"
  >
    <div
      class="avatar"
    >
      Identicon
    </div>
    <div
      class="nameLine"
    >
      <span
        class="name"
      >
        Light DID Identity
      </span>
      <div
        class="wrapper"
      >
        <button
          aria-expanded="false"
          aria-haspopup="true"
          aria-label="Identity options"
          class="toggle"
          role="button"
          tabindex="0"
          title="Identity options"
          type="button"
        />
      </div>
    </div>
  </section>
</div>
`;

exports[`IdentitySlide should render 1`] = `
<div>
  <section
    class="container"
  >
    <div
      class="avatar"
    >
      Identicon
    </div>
    <div
      class="centeredNameLine"
    >
      <span
        class="name"
      >
        Light DID Identity
      </span>
    </div>
  </section>
</div>
`;

exports[`IdentitySlideNew should render 1`] = `
<div>
  <section>
    <a
      aria-label="Add Identity"
      class="add"
      href="/identity/add"
      title="Add Identity"
    />
  </section>
</div>
`;
